src/repr/src/row/encoding.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A permanent storage encoding for rows.
+//!
+//! To minimize cycles spent encoding/decoding, we use Row's internal storage
+//! format, with a prefix to allow for migrations if the internal format
+//! changes.
+//!
+//! ```none
+//! <version: 0u8><v0_encoding: &[u8]>
+//! - Currently the only value for `version` is `0u8`.
+//!
+//! <v0_encoding> is <len_bytes: u64><row_data: &[u8]>
+//! - `len_bytes` is little endian.
+//! - `row_data` is `len_bytes` long.
+//!
+//! <row_data> is repeated <tag>(<datum: &[u8]>)?
+//! - `tag` is one of the following values and determines the presence, length,
+//!   and encoding of `datum`.
+//! - All `datum` integers, signed and unsigned, are little endian.
+//!
+//! <tag=0u8> is Datum::Null
+//! <tag=1u8> is Datum::False
+//! <tag=2u8> is Datum::True
+//! <tag=3u8><val: i32> is Datum::I32(val)
+//! WIP finish this
+//! ```
+
+use std::io::Read;
+
+use crate::Row;
+
+impl Row {
+    const CURRENT_VERSION: u8 = 0u8;
+
+    /// Encodes a row into the permanent storage format.
+    ///
+    /// This perfectly round-trips through [Row::decode]. It's guaranteed to be
+    /// readable by future versions of Materialize through v(TODO: Figure out
+    /// our policy).
+    pub fn encode<E: Extend<u8>>(&self, buf: &mut E) {
+        // WIP: All these .iter().copied() calls seem wrong
```

Comment:
Any idea what I'm doing wrong here?

Reply:
This could obviously take a `&mut Vec<u8>` but that would let it mess with earlier bytes that ideally it couldn't touch.

Reply:
So I could remove two of those `.iter().copied()` calls relatively easily

```
diff --git a/src/repr/src/row/encoding.rs b/src/repr/src/row/encoding.rs
index d7a2ef563..5aab52fd1 100644
--- a/src/repr/src/row/encoding.rs
+++ b/src/repr/src/row/encoding.rs
@@ -47,9 +47,9 @@ impl Row {
     /// our policy).
     pub fn encode<E: Extend<u8>>(&self, buf: &mut E) {
         // WIP: All these .iter().copied() calls seem wrong
-        buf.extend([Self::CURRENT_VERSION].iter().copied());
+        buf.extend([Self::CURRENT_VERSION]);
         let len_raw = self.data.len().to_le_bytes();
-        buf.extend(len_raw[..].iter().copied());
+        buf.extend(len_raw);
         buf.extend(self.data[..].iter().copied());
     }
```

I think you maybe want to use the extend impl over `&'a T` instead of `T` which seems like its designed to copy things out of slices based on the comment here https://doc.rust-lang.org/std/iter/trait.Extend.html#impl-Extend%3C%26%27a%20T%3E-4 but when I tried it I couldn't convince the compiler that `len_raw` lived long enough. Maybe I'm still grossly misunderstanding something :)

Reply:
The types you want are extendable with both &u8 and u8, so you could just specify that in the type signature:

```rust
    pub fn encode<E: Extend<u8> + for<'a> Extend<&'a u8>>(&self, buf: &mut E) {
        buf.extend([Self::CURRENT_VERSION]);
        buf.extend(self.data.len().to_le_bytes());
        buf.extend(&self.data);
    }
```

But the way you wrote it is good too!